src/Animal.js

```diff
@@ -0,0 +1,109 @@
+import { Organism } from './Organism';
+import { findRandomTileInArray } from './findRandomTileInArray';
+import { Berry } from './berry/Berry';
+
+export class Animal extends Organism {
+  constructor(board, startParameters) {
+    super(board, startParameters);
+    this.initiative = startParameters.initiative;
+    this.numberOfSteps = startParameters.numberOfSteps;
+    this.strength = startParameters.strength;
+  }
+  action = () => {
+    if (this.isAlive) {
+      // console.log(`My turn!`)
+      // console.log(this)
+      // console.log(this.isAlive)
+      const newTile = this.findTileForMove();
+      const opponent = newTile.currentOrganism;
+      if (opponent) {
+        if (this.shouldEat(opponent)) {
+          this.eatPlant(opponent, newTile);
+        } else if (this.shouldMate(opponent)) {
+          this.mate(opponent);
+        } else if (this.shouldFight(opponent)) {
+          this.fight(opponent);
+        }
+      } else if (this.shouldMove(newTile)) {
+        this.move(newTile);
+      }
+    }
+    // console.log('action: animal');
+    super.action();
+  };
+  findTileForMove = () => {
+    const nearestTiles = this.board.findNearestTiles(this);
+    // console.log(nearestTiles);
+    return findRandomTileInArray(nearestTiles);
+    // console.log(newTile);
+  };
+  move = (newTile) => {
+    // console.log(this.currentTile)
+    // console.log(newTile)
+    this.currentTile.clearTile();
+    newTile.addCurrentOrganism(this);
+    // console.log(newTile.currentOrganism)
+    // console.log(`I moved!`);
+  };
+
+  shouldMove = (newTile) => {
+    return !newTile.currentOrganism;
+  };
+  shouldEat = (opponent) => {
+    return !(opponent instanceof Animal);
+  };
+  eatPlant = (opponent, newTile) => {
+    newTile.killCurrentOrganism(this);
+    if (!(opponent instanceof Berry)) {
+      newTile.addCurrentOrganism(this);
+    } else {
+      this.currentTile.killCurrentOrganism(this);
+    }
+    // console.log(`I ate a plant`);
+  };
+  shouldMate = (opponent) => {
+    if (opponent.isAlive) {
+      return opponent instanceof this.constructor;
+    }
+  };
+  mate = (opponent) => {
+    const surroundingEmptyTiles = this.board.findEmptyTilesToMate(
+      this,
+      opponent,
+    );
+    // console.log(`I will mate`);
+    if (surroundingEmptyTiles.length > 0) {
```

Comment:
This should be a part of `shouldMate` method not the `mate` method itself

Reply:
hmm.. started to move the if statement to shouldMove(), but it seems to complicate the code even more. While I'm checking if there's a space for the child, I simultaneously pick the tile and use it later in mate(). If I move this to shouldMate(), I'll need to save picked tile somewhere to pass it to another function or look for the tile again. Any hints how to make it work?

Reply:
Alright let's leave it as is then ;)